--- input.pretty.loop
+++ optimized.loop
@@ -1,10 +1,10 @@
-for i0 in range(0, 10000) {
-  for i1 in range(0, 10000) {
-    B[i0][i1] = (i0 + i1);
+for i1 in range(0, 10000) {
+  for i2 in range(0, 10000) {
+    B[i1][i2] = (i1 + i2);
   }
 }
-for i0 in range(0, 10000) {
-  for i1 in range(0, 10000) {
-    C[i0][i1] = (B[(i0 - 1)][(i1 - 1)] + B[i0][(i1 - 1)]);
+for i1 in range(0, 10000) {
+  for i2 in range(0, 10000) {
+    C[i1][i2] = (B[(i1 - 1)][(i2 - 1)] + B[i1][(i2 - 1)]);
   }
 }
