--- input.pretty.loop
+++ optimized.loop
@@ -1,30 +1,157 @@
 context(ny, nx, nz);
 
-for i0 in range(4, (((ny + 9) + -2) + 1)) {
-  for i1 in range(4, (((nx + 9) + -3) + 1)) {
-    for i2 in range(4, (((nz + 9) + -3) + 1)) {
-      ab[i0][i1][i2] = (((((f60 * (a[(i0 - 1)][i1][i2] + a[i0][i1][i2])) + (f61 * (a[(i0 - 2)][i1][i2] + a[(i0 + 1)][i1][i2]))) + (f62 * (a[(i0 - 3)][i1][i2] + a[(i0 + 2)][i1][i2]))) * thirddtbydy) * uyb[i0][i1][i2]);
+if ((-2 <= nz && (-2 <= nx && -2 <= ny))) {
+  for i0 in range(4, (nx + 7)) {
+    for i1 in range(4, (ny + 7)) {
+      ab[i1][i0][4] = (((((f60 * (a[(i1 - 1)][i0][4] + a[i1][i0][4])) + (f61 * (a[(i1 - 2)][i0][4] + a[(i1 + 1)][i0][4]))) + (f62 * (a[(i1 - 3)][i0][4] + a[(i1 + 2)][i0][4]))) * thirddtbydy) * uyb[i1][i0][4]);
+      al[i1][i0][4] = (((((f60 * (a[i1][(i0 - 1)][4] + a[i1][i0][4])) + (f61 * (a[i1][(i0 - 2)][4] + a[i1][(i0 + 1)][4]))) + (f62 * (a[i1][(i0 - 3)][4] + a[i1][(i0 + 2)][4]))) * thirddtbydx) * uxl[i1][i0][4]);
+      af[i1][i0][4] = (((((f60 * (a[i1][i0][(4 - 1)] + a[i1][i0][4])) + (f61 * (a[i1][i0][(4 - 2)] + a[i1][i0][(4 + 1)]))) + (f62 * (a[i1][i0][(4 - 3)] + a[i1][i0][(4 + 2)]))) * thirddtbydz) * uzf[i1][i0][4]);
+    }
+    for i1 in range((ny + 7), (ny + 8)) {
+      for i2 in range((ny + 7), (ny + 8)) {
+        ab[i2][i0][4] = (((((f60 * (a[(i2 - 1)][i0][4] + a[i2][i0][4])) + (f61 * (a[(i2 - 2)][i0][4] + a[(i2 + 1)][i0][4]))) + (f62 * (a[(i2 - 3)][i0][4] + a[(i2 + 2)][i0][4]))) * thirddtbydy) * uyb[i2][i0][4]);
+      }
+    }
+  }
+  for i0 in range((nx + 7), (nx + 8)) {
+    for i1 in range(4, (ny + 7)) {
+      al[i1][i0][4] = (((((f60 * (a[i1][(i0 - 1)][4] + a[i1][i0][4])) + (f61 * (a[i1][(i0 - 2)][4] + a[i1][(i0 + 1)][4]))) + (f62 * (a[i1][(i0 - 3)][4] + a[i1][(i0 + 2)][4]))) * thirddtbydx) * uxl[i1][i0][4]);
     }
   }
 }
-for i0 in range(4, (((ny + 9) + -3) + 1)) {
-  for i1 in range(4, (((nx + 9) + -2) + 1)) {
-    for i2 in range(4, (((nz + 9) + -3) + 1)) {
-      al[i0][i1][i2] = (((((f60 * (a[i0][(i1 - 1)][i2] + a[i0][i1][i2])) + (f61 * (a[i0][(i1 - 2)][i2] + a[i0][(i1 + 1)][i2]))) + (f62 * (a[i0][(i1 - 3)][i2] + a[i0][(i1 + 2)][i2]))) * thirddtbydx) * uxl[i0][i1][i2]);
+if ((-2 <= nx && -2 <= ny)) {
+  for i0 in range(5, (nz + 7)) {
+    for i1 in range(4, (ny + 7)) {
+      ab[i1][4][i0] = (((((f60 * (a[(i1 - 1)][4][i0] + a[i1][4][i0])) + (f61 * (a[(i1 - 2)][4][i0] + a[(i1 + 1)][4][i0]))) + (f62 * (a[(i1 - 3)][4][i0] + a[(i1 + 2)][4][i0]))) * thirddtbydy) * uyb[i1][4][i0]);
+      al[i1][4][i0] = (((((f60 * (a[i1][(4 - 1)][i0] + a[i1][4][i0])) + (f61 * (a[i1][(4 - 2)][i0] + a[i1][(4 + 1)][i0]))) + (f62 * (a[i1][(4 - 3)][i0] + a[i1][(4 + 2)][i0]))) * thirddtbydx) * uxl[i1][4][i0]);
+      af[i1][4][i0] = (((((f60 * (a[i1][4][(i0 - 1)] + a[i1][4][i0])) + (f61 * (a[i1][4][(i0 - 2)] + a[i1][4][(i0 + 1)]))) + (f62 * (a[i1][4][(i0 - 3)] + a[i1][4][(i0 + 2)]))) * thirddtbydz) * uzf[i1][4][i0]);
+    }
+    for i1 in range((ny + 7), (ny + 8)) {
+      for i2 in range((ny + 7), (ny + 8)) {
+        ab[i2][4][i0] = (((((f60 * (a[(i2 - 1)][4][i0] + a[i2][4][i0])) + (f61 * (a[(i2 - 2)][4][i0] + a[(i2 + 1)][4][i0]))) + (f62 * (a[(i2 - 3)][4][i0] + a[(i2 + 2)][4][i0]))) * thirddtbydy) * uyb[i2][4][i0]);
+      }
+    }
+    for i1 in range(5, (nx + 7)) {
+      ab[4][i1][i0] = (((((f60 * (a[(4 - 1)][i1][i0] + a[4][i1][i0])) + (f61 * (a[(4 - 2)][i1][i0] + a[(4 + 1)][i1][i0]))) + (f62 * (a[(4 - 3)][i1][i0] + a[(4 + 2)][i1][i0]))) * thirddtbydy) * uyb[4][i1][i0]);
+      al[4][i1][i0] = (((((f60 * (a[4][(i1 - 1)][i0] + a[4][i1][i0])) + (f61 * (a[4][(i1 - 2)][i0] + a[4][(i1 + 1)][i0]))) + (f62 * (a[4][(i1 - 3)][i0] + a[4][(i1 + 2)][i0]))) * thirddtbydx) * uxl[4][i1][i0]);
+      af[4][i1][i0] = (((((f60 * (a[4][i1][(i0 - 1)] + a[4][i1][i0])) + (f61 * (a[4][i1][(i0 - 2)] + a[4][i1][(i0 + 1)]))) + (f62 * (a[4][i1][(i0 - 3)] + a[4][i1][(i0 + 2)]))) * thirddtbydz) * uzf[4][i1][i0]);
+      for i2 in range(5, (ny + 7)) {
+        for i3 in range((i2 + -1), i2) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+        ab[i2][i1][i0] = (((((f60 * (a[(i2 - 1)][i1][i0] + a[i2][i1][i0])) + (f61 * (a[(i2 - 2)][i1][i0] + a[(i2 + 1)][i1][i0]))) + (f62 * (a[(i2 - 3)][i1][i0] + a[(i2 + 2)][i1][i0]))) * thirddtbydy) * uyb[i2][i1][i0]);
+        al[i2][i1][i0] = (((((f60 * (a[i2][(i1 - 1)][i0] + a[i2][i1][i0])) + (f61 * (a[i2][(i1 - 2)][i0] + a[i2][(i1 + 1)][i0]))) + (f62 * (a[i2][(i1 - 3)][i0] + a[i2][(i1 + 2)][i0]))) * thirddtbydx) * uxl[i2][i1][i0]);
+        af[i2][i1][i0] = (((((f60 * (a[i2][i1][(i0 - 1)] + a[i2][i1][i0])) + (f61 * (a[i2][i1][(i0 - 2)] + a[i2][i1][(i0 + 1)]))) + (f62 * (a[i2][i1][(i0 - 3)] + a[i2][i1][(i0 + 2)]))) * thirddtbydz) * uzf[i2][i1][i0]);
+      }
+      for i2 in range((ny + 7), (ny + 8)) {
+        for i3 in range((ny + 6), (ny + 7)) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+        for i3 in range((ny + 7), (ny + 8)) {
+          ab[i3][i1][i0] = (((((f60 * (a[(i3 - 1)][i1][i0] + a[i3][i1][i0])) + (f61 * (a[(i3 - 2)][i1][i0] + a[(i3 + 1)][i1][i0]))) + (f62 * (a[(i3 - 3)][i1][i0] + a[(i3 + 2)][i1][i0]))) * thirddtbydy) * uyb[i3][i1][i0]);
+        }
+      }
+    }
+    for i1 in range((nx + 7), (nx + 8)) {
+      al[4][i1][i0] = (((((f60 * (a[4][(i1 - 1)][i0] + a[4][i1][i0])) + (f61 * (a[4][(i1 - 2)][i0] + a[4][(i1 + 1)][i0]))) + (f62 * (a[4][(i1 - 3)][i0] + a[4][(i1 + 2)][i0]))) * thirddtbydx) * uxl[4][i1][i0]);
+      for i2 in range(5, (ny + 7)) {
+        for i3 in range((i2 + -1), i2) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+        al[i2][i1][i0] = (((((f60 * (a[i2][(i1 - 1)][i0] + a[i2][i1][i0])) + (f61 * (a[i2][(i1 - 2)][i0] + a[i2][(i1 + 1)][i0]))) + (f62 * (a[i2][(i1 - 3)][i0] + a[i2][(i1 + 2)][i0]))) * thirddtbydx) * uxl[i2][i1][i0]);
+      }
+      for i2 in range((ny + 7), (ny + 8)) {
+        for i3 in range((ny + 6), (ny + 7)) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+      }
     }
   }
 }
-for i0 in range(4, (((ny + 9) + -3) + 1)) {
-  for i1 in range(4, (((nx + 9) + -3) + 1)) {
-    for i2 in range(4, (((nz + 9) + -2) + 1)) {
-      af[i0][i1][i2] = (((((f60 * (a[i0][i1][(i2 - 1)] + a[i0][i1][i2])) + (f61 * (a[i0][i1][(i2 - 2)] + a[i0][i1][(i2 + 1)]))) + (f62 * (a[i0][i1][(i2 - 3)] + a[i0][i1][(i2 + 2)]))) * thirddtbydz) * uzf[i0][i1][i2]);
+if ((-2 <= nz && (-2 <= nx && -2 <= ny))) {
+  for i0 in range((nz + 7), (nz + 8)) {
+    for i1 in range(4, (ny + 7)) {
+      af[i1][4][i0] = (((((f60 * (a[i1][4][(i0 - 1)] + a[i1][4][i0])) + (f61 * (a[i1][4][(i0 - 2)] + a[i1][4][(i0 + 1)]))) + (f62 * (a[i1][4][(i0 - 3)] + a[i1][4][(i0 + 2)]))) * thirddtbydz) * uzf[i1][4][i0]);
+    }
+    for i1 in range(5, (nx + 7)) {
+      af[4][i1][i0] = (((((f60 * (a[4][i1][(i0 - 1)] + a[4][i1][i0])) + (f61 * (a[4][i1][(i0 - 2)] + a[4][i1][(i0 + 1)]))) + (f62 * (a[4][i1][(i0 - 3)] + a[4][i1][(i0 + 2)]))) * thirddtbydz) * uzf[4][i1][i0]);
+      for i2 in range(5, (ny + 7)) {
+        for i3 in range((i2 + -1), i2) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+        af[i2][i1][i0] = (((((f60 * (a[i2][i1][(i0 - 1)] + a[i2][i1][i0])) + (f61 * (a[i2][i1][(i0 - 2)] + a[i2][i1][(i0 + 1)]))) + (f62 * (a[i2][i1][(i0 - 3)] + a[i2][i1][(i0 + 2)]))) * thirddtbydz) * uzf[i2][i1][i0]);
+      }
+      for i2 in range((ny + 7), (ny + 8)) {
+        for i3 in range((ny + 6), (ny + 7)) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+      }
+    }
+    for i1 in range((nx + 7), (nx + 8)) {
+      for i2 in range(5, (ny + 8)) {
+        for i3 in range((i2 + -1), i2) {
+          for i4 in range((i1 + -1), i1) {
+            for i5 in range((i0 + -1), i0) {
+              athird[i3][i4][i5] = (((a[i3][i4][i5] + (al[i3][(i4 + 1)][i5] - al[i3][i4][i5])) + (ab[(i3 + 1)][i4][i5] - ab[i3][i4][i5])) + (af[i3][i4][(i5 + 1)] - af[i3][i4][i5]));
+            }
+          }
+        }
+      }
     }
   }
 }
-for i0 in range(4, (((ny + 9) + -3) + 1)) {
-  for i1 in range(4, (((nx + 9) + -3) + 1)) {
-    for i2 in range(4, (((nz + 9) + -3) + 1)) {
-      athird[i0][i1][i2] = (((a[i0][i1][i2] + (al[i0][(i1 + 1)][i2] - al[i0][i1][i2])) + (ab[(i0 + 1)][i1][i2] - ab[i0][i1][i2])) + (af[i0][i1][(i2 + 1)] - af[i0][i1][i2]));
+if ((-2 <= nx && (-2 <= ny && (nz <= -3 && -3 <= nz)))) {
+  for i0 in range(4, (nx + 7)) {
+    for i1 in range(4, (ny + 7)) {
+      af[i1][i0][4] = (((((f60 * (a[i1][i0][(4 - 1)] + a[i1][i0][4])) + (f61 * (a[i1][i0][(4 - 2)] + a[i1][i0][(4 + 1)]))) + (f62 * (a[i1][i0][(4 - 3)] + a[i1][i0][(4 + 2)]))) * thirddtbydz) * uzf[i1][i0][4]);
     }
   }
 }
+if ((-2 <= ny && (-2 <= nz && (nx <= -3 && -3 <= nx)))) {
+  for i0 in range(4, (ny + 7)) {
+    al[i0][4][4] = (((((f60 * (a[i0][(4 - 1)][4] + a[i0][4][4])) + (f61 * (a[i0][(4 - 2)][4] + a[i0][(4 + 1)][4]))) + (f62 * (a[i0][(4 - 3)][4] + a[i0][(4 + 2)][4]))) * thirddtbydx) * uxl[i0][4][4]);
+  }
+}
+if ((-2 <= nx && (-2 <= nz && (ny <= -3 && -3 <= ny)))) {
+  for i0 in range(4, (nx + 7)) {
+    ab[4][i0][4] = (((((f60 * (a[(4 - 1)][i0][4] + a[4][i0][4])) + (f61 * (a[(4 - 2)][i0][4] + a[(4 + 1)][i0][4]))) + (f62 * (a[(4 - 3)][i0][4] + a[(4 + 2)][i0][4]))) * thirddtbydy) * uyb[4][i0][4]);
+  }
+}
+if ((-2 <= ny && (nx <= -3 && -3 <= nx))) {
+  for i0 in range(5, (nz + 7)) {
+    for i1 in range(4, (ny + 7)) {
+      al[i1][4][i0] = (((((f60 * (a[i1][(4 - 1)][i0] + a[i1][4][i0])) + (f61 * (a[i1][(4 - 2)][i0] + a[i1][(4 + 1)][i0]))) + (f62 * (a[i1][(4 - 3)][i0] + a[i1][(4 + 2)][i0]))) * thirddtbydx) * uxl[i1][4][i0]);
+    }
+  }
+}
+if ((-2 <= nx && (ny <= -3 && -3 <= ny))) {
+  for i0 in range(5, (nz + 7)) {
+    for i1 in range(4, (nx + 7)) {
+      ab[4][i1][i0] = (((((f60 * (a[(4 - 1)][i1][i0] + a[4][i1][i0])) + (f61 * (a[(4 - 2)][i1][i0] + a[(4 + 1)][i1][i0]))) + (f62 * (a[(4 - 3)][i1][i0] + a[(4 + 2)][i1][i0]))) * thirddtbydy) * uyb[4][i1][i0]);
+    }
+  }
+}
