--- input.pretty.loop
+++ optimized.loop
@@ -1,30 +1,38 @@
 context(ny, nx, nz);
 
-for i0 in range(4, (ny + 8)) {
-  for i1 in range(4, (nx + 7)) {
-    for i2 in range(4, (nz + 7)) {
-      ab[i0][i1][i2] = (((((f60 * (a[(i0 - 1)][i1][i2] + a[i0][i1][i2])) + (f61 * (a[(i0 - 2)][i1][i2] + a[(i0 + 1)][i1][i2]))) + (f62 * (a[(i0 - 3)][i1][i2] + a[(i0 + 2)][i1][i2]))) * thirddtbydy) * uyb[i0][i1][i2]);
+if ((-2 <= nz && (-2 <= nx && -3 <= ny))) {
+  for i1 in range(4, (nz + 7)) {
+    for i2 in range(4, (nx + 7)) {
+      for i4 in range(4, (ny + 8)) {
+        ab[i4][i2][i1] = (((((f60 * (a[(i4 - 1)][i2][i1] + a[i4][i2][i1])) + (f61 * (a[(i4 - 2)][i2][i1] + a[(i4 + 1)][i2][i1]))) + (f62 * (a[(i4 - 3)][i2][i1] + a[(i4 + 2)][i2][i1]))) * thirddtbydy) * uyb[i4][i2][i1]);
+      }
     }
   }
 }
-for i0 in range(4, (ny + 7)) {
-  for i1 in range(4, (nx + 8)) {
-    for i2 in range(4, (nz + 7)) {
-      al[i0][i1][i2] = (((((f60 * (a[i0][(i1 - 1)][i2] + a[i0][i1][i2])) + (f61 * (a[i0][(i1 - 2)][i2] + a[i0][(i1 + 1)][i2]))) + (f62 * (a[i0][(i1 - 3)][i2] + a[i0][(i1 + 2)][i2]))) * thirddtbydx) * uxl[i0][i1][i2]);
+if ((-2 <= nz && (-3 <= nx && -2 <= ny))) {
+  for i1 in range(4, (ny + 7)) {
+    for i2 in range(4, (nx + 8)) {
+      for i3 in range(4, (nz + 7)) {
+        al[i1][i2][i3] = (((((f60 * (a[i1][(i2 - 1)][i3] + a[i1][i2][i3])) + (f61 * (a[i1][(i2 - 2)][i3] + a[i1][(i2 + 1)][i3]))) + (f62 * (a[i1][(i2 - 3)][i3] + a[i1][(i2 + 2)][i3]))) * thirddtbydx) * uxl[i1][i2][i3]);
+      }
     }
   }
 }
-for i0 in range(4, (ny + 7)) {
-  for i1 in range(4, (nx + 7)) {
-    for i2 in range(4, (nz + 8)) {
-      af[i0][i1][i2] = (((((f60 * (a[i0][i1][(i2 - 1)] + a[i0][i1][i2])) + (f61 * (a[i0][i1][(i2 - 2)] + a[i0][i1][(i2 + 1)]))) + (f62 * (a[i0][i1][(i2 - 3)] + a[i0][i1][(i2 + 2)]))) * thirddtbydz) * uzf[i0][i1][i2]);
+if ((-3 <= nz && (-2 <= nx && -2 <= ny))) {
+  for i1 in range(4, (ny + 7)) {
+    for i2 in range(4, (nx + 7)) {
+      for i3 in range(4, (nz + 8)) {
+        af[i1][i2][i3] = (((((f60 * (a[i1][i2][(i3 - 1)] + a[i1][i2][i3])) + (f61 * (a[i1][i2][(i3 - 2)] + a[i1][i2][(i3 + 1)]))) + (f62 * (a[i1][i2][(i3 - 3)] + a[i1][i2][(i3 + 2)]))) * thirddtbydz) * uzf[i1][i2][i3]);
+      }
     }
   }
 }
-for i0 in range(4, (ny + 7)) {
-  for i1 in range(4, (nx + 7)) {
-    for i2 in range(4, (nz + 7)) {
-      athird[i0][i1][i2] = (((a[i0][i1][i2] + (al[i0][(i1 + 1)][i2] - al[i0][i1][i2])) + (ab[(i0 + 1)][i1][i2] - ab[i0][i1][i2])) + (af[i0][i1][(i2 + 1)] - af[i0][i1][i2]));
+if ((-2 <= nz && (-2 <= nx && -2 <= ny))) {
+  for i1 in range(4, (ny + 7)) {
+    for i2 in range(4, (nx + 7)) {
+      for i3 in range(4, (nz + 7)) {
+        athird[i1][i2][i3] = (((a[i1][i2][i3] + (al[i1][(i2 + 1)][i3] - al[i1][i2][i3])) + (ab[(i1 + 1)][i2][i3] - ab[i1][i2][i3])) + (af[i1][i2][(i3 + 1)] - af[i1][i2][i3]));
+      }
     }
   }
 }
