--- input.pretty.loop
+++ optimized.loop
@@ -1,11 +1,17 @@
 context(N, M, L);
 
-for i0 in range(0, N) {
-  a[i0] = b[i0];
+if (1 <= N) {
+  for i1 in range(0, N) {
+    a[i1] = b[i1];
+  }
 }
-for i0 in range(0, M) {
-  a[i0] = b[i0];
+if (1 <= M) {
+  for i1 in range(0, M) {
+    a[i1] = b[i1];
+  }
 }
-for i0 in range(0, L) {
-  a[i0] = b[i0];
+if (1 <= L) {
+  for i1 in range(0, L) {
+    a[i1] = b[i1];
+  }
 }
