--- input.pretty.loop
+++ optimized.loop
@@ -1,10 +1,71 @@
 context(M, N);
 
-for i0 in range(1, (M + 1)) {
-  for i1 in range(i0, (M + 1)) {
-    for i2 in range(1, (N + 1)) {
-      symmat[i0][i1] = (symmat[i0][i1] + (data[i2][i0] * data[i2][i1]));
+for i0 in range(1, min((1 + M), (1 + N))) {
+  for i1 in range(1, (1 + (i0 + -1))) {
+    for i2 in range(0, (0 + 1)) {
+      for i3 in range(i1, (1 + M)) {
+        for i4 in range(i1, (i1 + 1)) {
+          for i5 in range(i3, (i3 + 1)) {
+            for i6 in range(i0, (i0 + 1)) {
+              symmat[i4][i5] = (symmat[i4][i5] + (data[i6][i4] * data[i6][i5]));
+            }
+          }
+        }
+      }
     }
-    symmat[i1][i0] = symmat[i0][i1];
+  }
+  for i1 in range(i0, (1 + M)) {
+    for i2 in range(0, (0 + 1)) {
+      for i3 in range(0, (0 + 1)) {
+        for i4 in range(i0, (i0 + 1)) {
+          for i5 in range(i1, (i1 + 1)) {
+            for i6 in range(0, (0 + 1)) {
+              symmat[i5][i4] = symmat[i4][i5];
+            }
+          }
+        }
+      }
+      for i3 in range(i1, (1 + M)) {
+        for i4 in range(i1, (i1 + 1)) {
+          for i5 in range(i3, (i3 + 1)) {
+            for i6 in range(i0, (i0 + 1)) {
+              symmat[i4][i5] = (symmat[i4][i5] + (data[i6][i4] * data[i6][i5]));
+            }
+          }
+        }
+      }
+    }
   }
 }
+for i0 in range(max(1, (N + 1)), (1 + M)) {
+  for i1 in range(i0, (1 + M)) {
+    for i2 in range(0, (0 + 1)) {
+      for i3 in range(0, (0 + 1)) {
+        for i4 in range(i0, (i0 + 1)) {
+          for i5 in range(i1, (i1 + 1)) {
+            for i6 in range(0, (0 + 1)) {
+              symmat[i5][i4] = symmat[i4][i5];
+            }
+          }
+        }
+      }
+    }
+  }
+}
+if ((-1 * M) <= -1) {
+  for i0 in range((M + 1), (1 + N)) {
+    for i1 in range(1, (1 + M)) {
+      for i2 in range(0, (0 + 1)) {
+        for i3 in range(i1, (1 + M)) {
+          for i4 in range(i1, (i1 + 1)) {
+            for i5 in range(i3, (i3 + 1)) {
+              for i6 in range(i0, (i0 + 1)) {
+                symmat[i4][i5] = (symmat[i4][i5] + (data[i6][i4] * data[i6][i5]));
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
