--- input.pretty.loop
+++ optimized.loop
@@ -1,14 +1,50 @@
 context(T, N);
 
-for i0 in range(0, T) {
-  for i1 in range(2, (N + -1)) {
-    for i2 in range(2, (N + -1)) {
-      b[i1][i2] = (0.2 * ((((a[i1][i2] + a[i1][(i2 - 1)]) + a[i1][(1 + i2)]) + a[(1 + i1)][i2]) + a[(i1 - 1)][i2]));
+if (4 <= N) {
+  for i0 in range(0, T) {
+    for i1 in range(((2 * i0) + 2), ((2 * i0) + 3)) {
+      for i2 in range(((2 * i0) + 2), (N + ((2 * i0) + -1))) {
+        for i3 in range(((-2 * i0) + i2), ((-2 * i0) + (i2 + 1))) {
+          b[2][i3] = (0.2 * ((((a[2][i3] + a[2][(i3 - 1)]) + a[2][(1 + i3)]) + a[(1 + 2)][i3]) + a[(2 - 1)][i3]));
+        }
+      }
     }
-  }
-  for i1 in range(2, (N + -1)) {
-    for i2 in range(2, (N + -1)) {
-      a[i1][i2] = b[i1][i2];
+    for i1 in range(((2 * i0) + 3), (N + ((2 * i0) + -1))) {
+      for i2 in range(((2 * i0) + 2), ((2 * i0) + 3)) {
+        if (((i2 + -2) % 2) == 0) {
+          for i3 in range((i1 + ((-1 * i2) + 2)), (i1 + ((-1 * i2) + 3))) {
+            b[i3][2] = (0.2 * ((((a[i3][2] + a[i3][(2 - 1)]) + a[i3][(1 + 2)]) + a[(1 + i3)][2]) + a[(i3 - 1)][2]));
+          }
+        }
+      }
+      for i2 in range(((2 * i0) + 3), (N + ((2 * i0) + -1))) {
+        for i3 in range(((-2 * i0) + (i1 + -1)), ((-2 * i0) + i1)) {
+          for i4 in range(((-2 * i0) + (i2 + -1)), ((-2 * i0) + i2)) {
+            a[i3][i4] = b[i3][i4];
+          }
+        }
+        for i3 in range(((-2 * i0) + i1), ((-2 * i0) + (i1 + 1))) {
+          for i4 in range(((-2 * i0) + i2), ((-2 * i0) + (i2 + 1))) {
+            b[i3][i4] = (0.2 * ((((a[i3][i4] + a[i3][(i4 - 1)]) + a[i3][(1 + i4)]) + a[(1 + i3)][i4]) + a[(i3 - 1)][i4]));
+          }
+        }
+      }
+      for i2 in range((N + ((2 * i0) + -1)), (N + (2 * i0))) {
+        for i3 in range(((-2 * i0) + (i1 + -1)), ((-2 * i0) + i1)) {
+          for i4 in range(((-2 * i0) + (i2 + -1)), ((-2 * i0) + i2)) {
+            a[i3][i4] = b[i3][i4];
+          }
+        }
+      }
+    }
+    for i1 in range((N + ((2 * i0) + -1)), (N + (2 * i0))) {
+      for i2 in range(((2 * i0) + 3), (i1 + 1)) {
+        for i3 in range((N + -2), (N + -1)) {
+          for i4 in range(((-2 * i0) + (i2 + -1)), ((-2 * i0) + i2)) {
+            a[i3][i4] = b[i3][i4];
+          }
+        }
+      }
     }
   }
 }
