--- input.pretty.loop
+++ optimized.loop
@@ -1,11 +1,27 @@
 context(NMAX);
 
-for i0 in range(0, NMAX) {
-  for i1 in range(0, NMAX) {
-    for i2 in range(0, (i1 + -1)) {
-      c[i0][i2] = (c[i0][i2] + (a[i1][i2] * b[i0][i1]));
+if (3 <= NMAX) {
+  for i0 in range(0, NMAX) {
+    for i1 in range(2, NMAX) {
+      for i2 in range(0, (i1 + -1)) {
+        c[i0][i1] = (c[i0][i1] + (a[i1][i1] * b[i0][i1]));
+      }
+    }
+  }
+}
+if (1 <= NMAX) {
+  for i0 in range(0, NMAX) {
+    for i1 in range(0, NMAX) {
       c[i0][i1] = (c[i0][i1] + (a[i1][i1] * b[i0][i1]));
     }
-    c[i0][i1] = (c[i0][i1] + (a[i1][i1] * b[i0][i1]));
   }
 }
+if (3 <= NMAX) {
+  for i0 in range(0, NMAX) {
+    for i1 in range(2, NMAX) {
+      for i2 in range(0, (i1 + -1)) {
+        c[i0][i2] = (c[i0][i2] + (a[i1][i2] * b[i0][i1]));
+      }
+    }
+  }
+}
