--- input.pretty.loop
+++ optimized.loop
@@ -1,12 +1,10 @@
 context(N);
 
 for i0 in range(0, N) {
-  for i1 in range(0, N) {
-    y[i0] = (y[i0] + (a[i0][i1] * x[i1]));
+  for i2 in range(0, N) {
+    y[i0] = (y[i0] + (a[i0][i2] * x[i2]));
+  }
+  for i2 in range(0, N) {
+    z[i2] = (z[i2] + (b[i2][i0] * y[i0]));
   }
 }
-for i0 in range(0, N) {
-  for i1 in range(0, N) {
-    z[i0] = (z[i0] + (b[i0][i1] * y[i1]));
-  }
-}
