--- input.pretty.loop
+++ optimized.loop
@@ -1,9 +1,11 @@
 context(M, N, K);
 
-for i0 in range(0, M) {
-  for i1 in range(0, N) {
-    for i2 in range(0, K) {
-      C[i0][i1] = ((beta * C[i0][i1]) + ((alpha * A[i0][i2]) * B[i2][i1]));
+if ((1 <= M && 1 <= N)) {
+  for i0 in range(0, K) {
+    for i1 in range(0, N) {
+      for i3 in range(0, M) {
+        C[i3][i1] = ((beta * C[i3][i1]) + ((alpha * A[i3][i0]) * B[i0][i1]));
+      }
     }
   }
 }
