--- input.pretty.loop
+++ optimized.loop
@@ -1,9 +1,21 @@
 context(M, N, K);
 
-for i0 in range(0, M) {
-  for i1 in range(0, N) {
-    for i2 in range(0, K) {
-      C[i0][i1] = ((beta * C[i0][i1]) + ((alpha * A[i0][i2]) * B[i2][i1]));
+if (((-1 * M) <= -1 && (-1 * N) <= -1)) {
+  for i0 in range(0, (1 + (K + -1))) {
+    for i1 in range(0, (1 + (N + -1))) {
+      for i2 in range(0, (1 + (M + -1))) {
+        for i3 in range(i2, (i2 + 1)) {
+          for i4 in range(i1, (i1 + 1)) {
+            for i5 in range(i0, (i0 + 1)) {
+              for i6 in range(0, (0 + 1)) {
+                for i7 in range(0, (0 + 1)) {
+                  C[i3][i4] = ((beta * C[i3][i4]) + ((alpha * A[i3][i5]) * B[i5][i4]));
+                }
+              }
+            }
+          }
+        }
+      }
     }
   }
 }
