--- input.pretty.loop
+++ optimized.loop
@@ -1,10 +1,17 @@
-for i0 in range(0, 2) {
-  for i1 in range(0, 2) {
-    A[((2 * i0) + i1)] = 1;
+for i0 in range(0, 4) {
+  if (i0 <= 1) {
+    A[((2 * 0) + i0)] = 1;
+  }
+  if (2 <= i0) {
+    for i1 in range((i0 + -2), (i0 + -1)) {
+      A[((2 * 1) + i1)] = 1;
+    }
+  }
+  for i1 in range(max(((i0 + 2) / 2), 1), min(((i0 + 4) / 2), 3)) {
+    for i2 in range((i1 + -1), i1) {
+      for i3 in range((i0 + (-2 * i2)), (i0 + ((-2 * i2) + 1))) {
+        B[((2 * i2) + i3)] = A[((2 * i2) + i3)];
+      }
+    }
   }
 }
-for i0 in range(0, 2) {
-  for i1 in range(0, 2) {
-    B[((2 * i0) + i1)] = A[((2 * i0) + i1)];
-  }
-}
